Fix index undefined error for Customizer rendered stories

divyarajmasani opened the pull request:
## Context

This PR tries to resolve some of the issues found while integrating the theme support.

## Summary

    - Fixes undefined index notice in Renderer class when trying to use Customizer settings.
    - Adds title show/hide toggle customizer control for 'circles' view.
    - Replaces static options like 'order' and 'layouts' to use common utility functions.
    - Fixes issue where multiple carousels were not working on the same page.
    - Fixes issue for AMP where multiple lightbox of the same story was not working as expected.
    - Fix: Select story modal was not closing if no stories were selected.
    - Fix: Gutenberg warning for usage of custom block toolbar.
    - Fix: Circles carousel were not initialized properly on mobile devices.
    - Fix: Block styles were not loaded properly in editor.

## Relevant Technical Choices

- Uses a static value to keep updating the number of Renderer.php instances. This is to identify the number of active renderer classes for supporting multiple carousels/lightboxes.
- Could have used [spl_object_id()](https://www.php.net/manual/en/function.spl-object-id.php) for the same but it requires php 7 ( >= 7.2.0 )

## To-do

- Streamline embedding options for various mediums used i.e., shortcode, customizer, widget, block etc. As required.

## User-facing changes

- N/A

## Testing Instructions

- PHP notices issue.
   - Add theme support for the web-stories with `add_theme_support( 'web-stories' )` for your theme.
   - Render the stories in your theme's `header.php` with `new Customizer()->render_stories();`.
   - You'll see the notices on the screen if the debugger is on.
  
- Reproduction steps for carousel not working.
   - Add two or more carousel on the same page, ( shortcode, block, customizer, widget any ).
   - Notice that the first carousel is working just fine, but the second carousel and others are not working as expected.

- Lightbox not working, steps to reproduce.
   - Add two or more Web Stories block/shortcode on the same page.
   - Ensure that at least one story is common among inserted web stories.
   - Observe by clicking the common story that it won't open the lightbox.

### QA

- Follow the above-mentioned testing instructions to validate the issue and confirm that the same has been fixed.

### UAT

<!--
Sometimes the testing instructions for UAT can differ from the ones for QA.
-->

<!-- ignore-task-list-start -->
- [ ] UAT should use the same steps as above.
<!-- ignore-task-list-end -->

<!--
If the above checkbox has not been checked, write down all steps necessary for user acceptance testing take to test this PR.
-->

## Reviews

### Does this PR have a security-related impact?

<!-- Examples: new APIs, changes to KSES, etc.  -->

### Does this PR change what data or activity we track or use?

<!-- Examples: changes to telemetry, new third-party APIs -->

### Does this PR have a legal-related impact?

<!-- Examples: new images with unknown sources, new production dependencies with incompatible licenses -->

---

spacedmonkey commented on includes/TinyMCE.php:
```diff
@@ -59,7 +59,9 @@ public function init() {
 	 */
 	public function tinymce_web_stories_button( array $buttons ) {
-		array_push( $buttons, 'web_stories' );
+		if ( ! $this->is_block_editor() ) {
```
Instead of having this in every function, why not just add in the init function.

```
	public function init() {
               if ( $this->is_block_editor() ) {
                     return;
               }
```

divyarajmasani replied:
`init` runs too early and `$current_screen` is not set there yet, hence put it for the individual actions ensuring that $current_screen will be there.

spacedmonkey replied:
init is loaded on `init` action. But there is no reason, we could load this on a different action, like `admin_init` or another action where screen is setup.  Doing this in 4 places isn't very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

spacedmonkey replied:
We could even use `current_screen` action.

spacedmonkey commented on includes/Stories_Renderer/Renderer.php:
```diff
@@ -65,6 +65,18 @@ abstract class Renderer implements RenderingInterface, Iterator {
+	 *
+	 * @var int
+	 */
+	public static $obj_id = 0;
```
Why is this static? It is never referenced statically.

spacedmonkey replied:
It also doesn't need to be public. It is only ever outputted to `get_obj_id`.

divyarajmasani replied:
As mentioned in the PR description, we needed to identify different instances of the rendered web stories containers, for that have kept a static property which can be accessed.

You right maybe since we don't need to access it publicly, we don't need it to be public.

spacedmonkey commented on includes/Stories_Renderer/Renderer.php:
```diff
@@ -129,12 +141,21 @@ abstract class Renderer implements RenderingInterface, Iterator {
+	 *
+	 * @return int Object id.
+	 */
+	public static function get_obj_id() {
```
Same point as approve, why does this need to be static?

swissspidy commented on tests/phpunit/tests/Story_Query.php:
```diff
@@ -78,7 +78,9 @@ public static function wpSetUpBeforeClass( $factory ) {
 			'show_date'                 => false,
+			'show_excerpt'              => false,
 			'show_stories_archive_link' => false,
+			'has_square_corners'        => false,
```
What is the `has_square_corners` option used for / where is it used?

divyarajmasani replied:
This option defines whether to show curved corners or sharp one for the stories. Every 'renderer' instance is created and passed this `Story_Query` object. so you can consider this as defaults that will get passed to the `Renderer`.
And in turn 'Renderer' add a class if it is set to true.

swissspidy replied:
In which Renderer is `has_square_corners` `true? Just so I can take a look.

Is this an actual option in the UI?

divyarajmasani replied:
Check the block inspector controls 'Show sharp corners', here
> https://google-webstories.rt.gw/wp-admin/post.php?post=1099&action=edit
This toggle option is visible for all view types except `circles`.

swissspidy commented on includes/Stories_Renderer/Renderer.php:
```diff
@@ -65,6 +65,18 @@ abstract class Renderer implements RenderingInterface, Iterator {
 	 */
 	const LIGHTBOX_SCRIPT_HANDLE = 'lightbox';
 
+	/**
+	 * Object ID for the Renderer class.
+	 * To enable support for multiple carousels and lightboxes
+	 * on the same page, we needed to identify each Renderer instance.
+	 *
+	 * This variable is used to add appropriate class to the Web Stories
+	 * wrapper.
+	 *
+	 * @var int
+	 */
+	protected static $obj_id = 0;
```
What if I use the same instance to render multiple carousels? e.g. (pseudo code):

```php
$renderer =  new Renderer();
$renderer->render();
$renderer->render();
```

This prints two carousels, and clicking on one Glider arrow button will control both carousels.

Could we just have a static `$instance_id` variable within the render method  instead  of this class property?

divyarajmasani replied:
Yep, you are right if the same instance is used multiple times on a page to render the stories it will have the same id. 
Didn't think of this use case or we didn't design for such case. 

Still you would want me to move this $obj_id to the method and increment if as the `->render()` is called. Will update this.

swissspidy replied:
Yeah I'd just use a static `$instance_id` variable in `render()` that we can increment.

Thanks!

divyarajmasani replied:
Since our `render()` method in the `Renderer::class` doesn't actually render any HTML, but the child classes override the same to render the HTML. Now if we want to have this static variable at function level then we either have to define it in the `Renderer::class->render()` and then return it so that the child classes can actually access it. or have child classes add that variable statical ( not DRY ).

How about we keep it as is and move the increment to `render()` this will keep it where it should be and not required much strings to pull. thoughts? @swissspidy

swissspidy replied:
>  How about we keep it as is and move the increment to render() this will keep it where it should be and not required much strings to pull.

Let's do this for now then  👍  Just one suggestion to use `instance_id` instead of `obj_id`, seems more descriptive.

swissspidy commented on assets/src/web-stories-block/block/components/blockTypeSwitcher.js:
```diff
@@ -33,18 +33,25 @@ import { BLOCK_TYPES } from '../constants';
-        };
-      })}
-    />
+    <Toolbar label={__('Change Block Type', 'web-stories')}>
+      <ToolbarItem>
+        {(toolbarItemHTMLProps) => (
+          <DropdownMenu
+            icon={update}
+            toggleProps={toolbarItemHTMLProps}
+            label={__('Change block type dropdown', 'web-stories')}
+            controls={BLOCK_TYPES.filter(
+              (blockType) => blockType.id !== selectedBlockType
+            ).map((blockType) => {
+              return {
+                title: blockType.label,
+                onClick: () => setAttributes({ blockType: blockType.id }),
+              };
+            })}
+          />
+        )}
+      </ToolbarItem>
+    </Toolbar>
```
Looking at the [documentation](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/toolbar#development-guidelines), `<Toolbar label={__('Options', 'web-stories')` and `<DropdownMenu label={__('Block Type', 'web-stories')}` seems more accurate.

----


Aside:

Is there another word we could use instead of "block type"? Block types in Gutenberg are something different (`registerBlockType`, etc.), so this is a bit confusing.

divyarajmasani replied:
I'll update the labels.

For the `block type`, kept as is as these were supposed to be three separate block that we have merged together. Couldn't find more accurate name for this one. any suggestion?

swissspidy replied:
I have a hard time finding a good name as well.

Some ideas off the top of my head:

* contentType
* blockSubType
* blockContentType

divyarajmasani replied:
let's go with the `blockContentType`, this signifies what type of content the block has, is it picked by editors ( selected ) or latests ( general story river ).

divyarajmasani replied:
@swissspidy do we want to update the attributes and used variable names from `blockType` to `blockSubType` across the block implementation right.

swissspidy commented on includes/functions.php:
```diff
@@ -167,3 +168,19 @@ function get_stories_order() {
 	);
 }
+
+/**
+ * Renders lightbox markup.
+ *
+ * @return void
+ */
+function render_stories_lightboxes() {
+
+	if ( is_amp() ) {
+		Renderer::render_stories_with_lightbox_amp();
+	} else {
+		Renderer::render_stories_with_lightbox_noamp();
+	}
+}
+
+add_action( 'wp_footer', 'Google\Web_Stories\render_stories_lightboxes' );
```
Could you elaborate a bit on the reason for this change around `render_stories_lightboxes` and making `render_stories_with_lightbox_noamp`/`render_stories_with_lightbox_amp` static?

This doesn't seem like a very developer-friendly API as it makes it hard to grasp what is actually going on.

* `render_stories_lightboxes` is a function for seemingly no reason.
* `render_stories_lightboxes` calls two static methods which rely on the `Renderer` already having generated some output.
* There is no context on why this is being done outside `Renderer`

cc @spacedmonkey WDYT?

divyarajmasani replied:
I've have questions for this as well, so while I was adding style changes for the Core Themes, I found that most issues are relating to the `z-index` or stacking context. 
Since we were rendering `lightboxes` in the `web-stories-list` wrapper, sometimes other elements were showing on top of the stories lightbox. For example for `twenty-fourteen`, I kept the stories in sidebar and enabled the customizer stories, the lightbox for the sidebar stories has header stories on top of it.

This change mitigates that issue as we move our lightbox into top most context of the page. But I didn't know where to put this piece of code, need your suggestion.

spacedmonkey replied:
What is this doing and why is it rendered on every page? it seems super wasteful to render it everywhere, even if it is not needed.

spacedmonkey replied:
Functions are designed to part of the public api, this should live in a class and should be conditionally loaded.

spacedmonkey commented on includes/Stories_Renderer/Generic_Renderer.php:
```diff
@@ -89,12 +89,6 @@ public function render( array $args = [] ) {
 			</div>
 			<?php
 			$this->maybe_render_archive_link();
-
-			if ( ! $this->is_amp_request() ) {
-				$this->render_stories_with_lightbox_noamp();
-			} else {
-				$this->render_stories_with_lightbox_amp();
-			}
```
Why not have another method here. 

```
public function render_footer(){
    if ( ! $this->is_amp_request() ) {
		$this->render_stories_with_lightbox_noamp();
    } else {
		$this->render_stories_with_lightbox_amp();
   }
}
```

Then you can call 
```
$this->maybe_render_archive_link();
add_action( 'wp_footer', [ $this, 'render_footer' ]);
```

spacedmonkey commented on includes/Stories_Renderer/Renderer.php:
```diff
@@ -462,8 +486,11 @@ public function render_single_story_content() {
-		// Web Stories Styles for AMP and non-AMP pages.
-		wp_enqueue_style( self::STYLE_HANDLE );
+		// No need to load these styles on admin as editor styles are being loaded by the block.
+		if ( ! is_admin() ) {
+			// Web Stories Styles for AMP and non-AMP pages.
+			wp_enqueue_style( self::STYLE_HANDLE );
+		}
```
Can you explain this change a little more. I find it confusing

divyarajmasani replied:
As mentioned in the comment, the styles needed for the block to render properly on the editor is already being loaded by the Block class. 
This was adding duplicate styles on the admin screens as well. And we need this stylesheet to only load on front end.

And since this gets loaded on render, when the user first inserts the block they were seeing the broken block previously, hence imported front end style in editor styles for the block.

spacedmonkey commented on includes/Stories_Renderer/Renderer.php:
```diff
@@ -86,12 +93,19 @@ abstract class Renderer implements RenderingInterface, Iterator {
 	 * @var string A string of lightbox markup.
 	 */
-	protected $lightbox_html = '';
+	protected static $lightbox_html = '';
```
Why does this need to be static?

spacedmonkey commented on includes/Stories_Renderer/Renderer.php:
```diff
@@ -523,20 +550,14 @@ protected function render_story_with_poster() {
+			} else {
+				$this->generate_amp_lightbox_html_amp( $story_data );
+			}
+			self::$stories_loaded[] = $story_data->get_id();
```
Why does this need to be static? This feels like a normal classic class property to me.

spacedmonkey reviewed (changes requested):
I don't like any of the use of static properties or methods here. Let's remove all those and see where we are.

divyarajmasani commented:
> I don't like any of the use of static properties or methods here. Let's remove all those and see where we are.

As mentioned above, this `static` property changes was done to render `lightbox` elements at the top level context, to minimise conflicts with theme styles.

But, I agree maybe `static` can be changed. since we can't add `wp_footer` in the same `Renderer` class as it instantiates multiple time. I had kept that action into `functions.php` temporarily.

divyarajmasani commented:
For the lightbox markup and `static` values:

- Collect lightbox markup while rendering the story items on the page and render it on `wp_footer`. **( Current )**
- Maybe just collect the storyIDs/objects that are loaded on the page and then separately render lightbox markup on `wp_footer`.

In either case I will create a utility class maybe, to track if the Web Stories are rendered on the page, check for customizer option, block in content etc., and render lightbox on `wp_footer`.  Also maybe we can move all instantiation of other web stories related services like TinyMCE, shortcode, Cutomizer from `Plugin.php`.

What are your thoughts?
@swissspidy @spacedmonkey

spacedmonkey commented:
Even in the example, static is not required. You could just do this. 

```
function render_stories_lightboxes() {
        $renderer = new Renderer();
	if ( is_amp() ) {
		$renderer->render_stories_with_lightbox_amp();
	} else {
		$renderer->render_stories_with_lightbox_noamp();
	}
}
```

spacedmonkey commented:
> Collect lightbox markup while rendering the story items on the page and render it on wp_footer. ( Current )
> Maybe just collect the storyIDs/objects that are loaded on the page and then separately render lightbox markup on wp_footer.

Can I confirm the problem we are trying to solve here? If you have a lightbox on the page, you need to have a unique id to setup it up? So if you had 5 different stories on the page, you need to setup 5 different light boxes and output setup into the footer. Is that correct? 

WIthout understanding the problem fully here, I can't really advice. I don't feel like the description gives enough detail for me to understand.

divyarajmasani commented:
I understand, I could have been more detailed.

For now there were following issues:
- Multiple instances of Carousels were not working.
- If there were duplicate stories then lightbox was not working as expected on AMP pages.
- I came across many issues where the lightbox was getting overlapped by other theme elements like the sidebar, sticky header ( theme-2014 ). To resolve that I moved the lightbox to render on `wp_footer`.

For the current solution, a static `$instance_id` property does resolve the first two issues. And to keep track of loaded stories so that we don't render duplicate lightboxes, added one for that. While these two I believe are needed, the other one which stores the lightbox markup can be replaced with a renderer function which iterates over `loaded stories`, which was added to save this loop through stories.

Let me know if we should connect over slack.

divyarajmasani commented:
@spacedmonkey @swissspidy , I've removed the static methods. 

Now we will have one lightbox wrapper per instance in the footer. Rest is same.

<img width="553" alt="Screenshot 2021-02-17 at 5 22 59 PM" src="https://user-images.githubusercontent.com/6906779/108200685-cf9fc000-7144-11eb-8a53-f02174eb2da7.png">

swissspidy commented on includes/Stories_Renderer/Renderer.php:
```diff
@@ -129,10 +144,18 @@ abstract class Renderer implements RenderingInterface, Iterator {
+	 *
+	 * @return int Object id.
+	 */
+	protected static function get_instance_id() {
```
Is this method used anywhere? Can't find references.

divyarajmasani replied:
not anymore, will have to remove. thanks.

swissspidy commented on assets/src/lightbox/web-stories-lightbox.js:
```diff
@@ -72,10 +66,14 @@ class Lightbox {
       card.addEventListener('click', (event) => {
         event.preventDefault();
-        this.player.show(this.stories[index].href);
+        const storyIdx = this.stories.findIndex(
+          (story) => story.href === card.dataset.storyUrl
+        );
+        this.player.show(this.stories[storyIdx].href);
```
```suggestion
        const storyObject = this.stories.find(
          (story) => story.href === card.dataset.storyUrl
        );
        this.player.show(storyObject.href);
```

swissspidy commented:
Thanks a lot @divyarajmasani!

Your explanation above is also very helpful and allows us to better understand what the code is actually trying to solve. It would be good to have this from the start :-)

I think we should just merge this for now. We can always iterate on it further once we've merged everything to our repo.

divyarajmasani commented:
Noted, 

I'm just testing few reported bugs with the existing PR. Though we still have to add specific styles for the Core themes but that should be a separate PR. I'll merge this one.
